src/lib/SX1280Driver/SX1280.cpp

```diff
@@ -0,0 +1,344 @@
+    double efeHz = 0.0;
+
+    efeRaw[0] = hal.ReadRegister(SX1280_REG_LR_ESTIMATED_FREQUENCY_ERROR_MSB);
+    efeRaw[1] = hal.ReadRegister(SX1280_REG_LR_ESTIMATED_FREQUENCY_ERROR_MSB + 1);
+    efeRaw[2] = hal.ReadRegister(SX1280_REG_LR_ESTIMATED_FREQUENCY_ERROR_MSB + 2);
+    efe = (efeRaw[0] << 16) | (efeRaw[1] << 8) | efeRaw[2];
+
+    efe &= SX1280_REG_LR_ESTIMATED_FREQUENCY_ERROR_MASK;
+
+    //efeHz = 1.55 * (double)complement2(efe, 20) / (1600.0 / (double)GetLoRaBandwidth() * 1000.0);
+    return efeHz;
+}
+
+void SX1280Driver::SetFIFOaddr(uint8_t txBaseAddr, uint8_t rxBaseAddr)
+{
+    uint8_t buf[2];
+
+    buf[0] = txBaseAddr;
+    buf[1] = rxBaseAddr;
+    hal.WriteCommand(SX1280_RADIO_SET_BUFFERBASEADDRESS, buf, 2);
+}
+
+void SX1280Driver::SetDioIrqParams(uint16_t irqMask, uint16_t dio1Mask, uint16_t dio2Mask, uint16_t dio3Mask)
+{
+    uint8_t buf[8];
+
+    buf[0] = (uint8_t)((irqMask >> 8) & 0x00FF);
+    buf[1] = (uint8_t)(irqMask & 0x00FF);
+    buf[2] = (uint8_t)((dio1Mask >> 8) & 0x00FF);
+    buf[3] = (uint8_t)(dio1Mask & 0x00FF);
+    buf[4] = (uint8_t)((dio2Mask >> 8) & 0x00FF);
+    buf[5] = (uint8_t)(dio2Mask & 0x00FF);
+    buf[6] = (uint8_t)((dio3Mask >> 8) & 0x00FF);
+    buf[7] = (uint8_t)(dio3Mask & 0x00FF);
+
+    hal.WriteCommand(SX1280_RADIO_SET_DIOIRQPARAMS, buf, 8);
+}
+
+void SX1280Driver::ClearIrqStatus(uint16_t irqMask)
+{
+    uint8_t buf[2];
+
+    buf[0] = (uint8_t)(((uint16_t)irqMask >> 8) & 0x00FF);
+    buf[1] = (uint8_t)((uint16_t)irqMask & 0x00FF);
+
+    hal.WriteCommand(SX1280_RADIO_CLR_IRQSTATUS, buf, 2);
+}
+
+void SX1280Driver::TXnbISR()
+{
+    //endTX = micros();
+    instance->ClearIrqStatus(SX1280_IRQ_RADIO_ALL);
+    instance->currOpmode = SX1280_MODE_FS; // radio goes to FS
+    //Serial.print("TOA: ");
+    //Serial.println(endTX - beginTX);
+    //instance->GetStatus();
+
+    // Serial.println("TXnbISR!");
+    //instance->GetStatus();
+    
+    //instance->GetStatus();
+    instance->TXdoneCallback();
+}
+
+void SX1280Driver::TXnb(volatile uint8_t *data, uint8_t length)
+{
+    instance->ClearIrqStatus(SX1280_IRQ_RADIO_ALL);
+    hal.TXenable(); // do first to allow PA stablise
+    instance->SetFIFOaddr(0x00, 0x00);   // not 100% sure if needed again
+    hal.WriteBuffer(0x00, data, length); //todo fix offset to equal fifo addr
+    instance->SetMode(SX1280_MODE_TX);
+    beginTX = micros();
+}
+
+void SX1280Driver::RXnbISR()
+{
+    instance->currOpmode = SX1280_MODE_FS;
+    instance->ClearIrqStatus(SX1280_IRQ_RADIO_ALL);
+    uint8_t FIFOaddr = instance->GetRxBufferAddr();
+    //hal.ReadBuffer(instance->GetRxBufferAddr(), instance->RXdataBuffer, 8);
+    hal.ReadBuffer(FIFOaddr, instance->RXdataBuffer, 8);
+    instance->RXdoneCallback();
+}
+
+void SX1280Driver::RXnb()
+{
+    //Serial.println("Start RX nb");
+    hal.setIRQassignment(SX1280_INTERRUPT_RX_DONE);
+    hal.RXenable();
+    instance->ClearIrqStatus(SX1280_IRQ_RADIO_ALL);
+    //instance->SetFIFOaddr(0x00, 0x00);
```

Comment:
Need this?

Reply:
hah, not sure! Unclear how the FIFO pointer works. I think it doesn't because I config the FIFO to wrap around automatically.